### PR TITLE
GCE L4 load balancer: enable migration of Instance Group management out of K/K. 

### PR DESCRIPTION
**Pull request by cezarygerard**
What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Cleanup Instance Group management in GCE L4 internal load balancer. 
Instance Group management has bern moved to k8s.io/ingress-gce controller.
Without this PR we will see bugs caused by controllers fighting over Instance Group management.
This PR allows GCE L4   internal load balancer to use more than one Instance Group, groups will be managed by k8s.io/ingress-gce controller.

#### Special notes for your reviewer:
Only affects GCE code

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go [hunk 1]
```diff
@@ -165,12 +165,35 @@ func TestEnsureInternalInstanceGroupsLimit(t *testing.T) {
...
+	for _, igName := range igNames {
+		ig := &compute.InstanceGroup{Name: igName}
+		err := gce.CreateInstanceGroup(ig, vals.ZoneName)
+		require.NoError(t, err)
+	}
+
+	igsFromCloud, err := gce.ensureInternalInstanceGroups(baseName, nodes)
+	require.NoError(t, err)
+	assert.Len(t, igsFromCloud, len(igNames), "Incorrect number of Instance Groups")
+}
```
Comment (cezarygerard) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal_test.go:195]: TODO:
add test case with IGs not to pick

**Comment by cezarygerard** [comment]
/assign @bowei 
/assign @freehan 
/assign @prameshj  
/assign @swetharepakula   

PTAL

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 2]
```diff
@@ -642,8 +652,10 @@ func (g *Cloud) ensureInternalInstanceGroupsDeleted(name string) error {
 
 	klog.V(2).Infof("ensureInternalInstanceGroupsDeleted(%v): attempting delete instance group in all %d zones", name, len(zones))
 	for _, z := range zones {
-		if err := g.DeleteInstanceGroup(name, z.Name); err != nil && !isNotFoundOrInUse(err) {
-			return err
+		if !g.AlphaFeatureGate.Enabled(AlphaFeatureNetLBRbs) {
```
Comment (swetharepakula) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:655]: Move the flag check to outside of the loop so we can skip the loop if enabled.

Comment (freehan) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:655]: add a comment to highlight that this is a reverse condition.

**Review by freehan — COMMENTED** [review]
LGTM overall
just nits

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 2 above)
Comment (cezarygerard) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: done

Comment (cezarygerard) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: added

**Comment by cezarygerard** [comment]
Review comments applied.

Who can give ok-to-test?

**Comment by cezarygerard** [comment]
@bowei @cheftako please take a look

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 3]
```diff
@@ -625,11 +625,21 @@ func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]s
 	klog.V(2).Infof("ensureInternalInstanceGroups(%v): %d nodes over %d zones in region %v", name, len(nodes), len(zonedNodes), g.region)
 	var igLinks []string
 	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
-		if err != nil {
-			return []string{}, err
+		if g.AlphaFeatureGate.Enabled(AlphaFeatureSkipIGsManagement) {
+			igs, err := g.FilterInstanceGroupsByName(name, zone)
+			if err != nil {
+				return []string{}, err
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:631]: can just return nil, err  (nil is the empty list)

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go [hunk 4]
```diff
@@ -625,11 +625,21 @@ func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]s
...
+			}
+			for _, ig := range igs {
+				igLinks = append(igLinks, ig.SelfLink)
+			}
+		} else {
+			igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+			if err != nil {
+				return []string{}, err
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:639]: same here

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instancegroup.go [hunk 5]
```diff
@@ -50,6 +50,16 @@ func (g *Cloud) DeleteInstanceGroup(name string, zone string) error {
 	return mc.Observe(g.c.InstanceGroups().Delete(ctx, meta.ZonalKey(name, zone)))
 }
 
+// FilterInstanceGroupsByName lists all InstanceGroups in the project and
+// zone that match the name regexp.
+func (g *Cloud) FilterInstanceGroupsByName(name, zone string) ([]*compute.InstanceGroup, error) {
+	ctx, cancel := cloud.ContextWithCallTimeout()
+	defer cancel()
+	mc := newInstanceGroupMetricContext("filter", zone)
+	v, err := g.c.InstanceGroups().List(ctx, zone, filter.Regexp("name", name))
```
Comment (bowei) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce_instancegroup.go:59]: did you check that List with regexp name works? Some of the cloud APIs don't understand the filter queries...

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 3 above)
Comment (cezarygerard) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: done

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_instancegroup.go (hunk 5 above)
Comment (cezarygerard) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_instancegroup.go]: fixed

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go (hunk 4 above)
Comment (cezarygerard) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go]: fixed

**Comment by cezarygerard** [comment]
@cheftako PTAL

this PR had an LGTM from Bowie until recent comments' fixes and rebase onto origin/master

**Comment by liggitt** [comment]
integration timeout flake tracked in https://github.com/kubernetes/kubernetes/issues/107857

**Comment by cheftako** [comment]
/assign @nckturner 
For cloud provider approval.

**Comment by nckturner** [comment]
@cezarygerard can you add some detail to the PR description?  The current description isn't totally clear if its a bug fix (only bug fixes are supposed to be merged into the legacy-cloud-providers). 

```
Cleanup instance group management in GCE L4 internal load balancer.
```

**Comment by cezarygerard** [comment]
@nckturner 
Updated the description
Since legacy-cloud-providers is closed for development so we have re-implemented Instance Group management in k8s.io/ingress-gce controller.
Without this PR we will see bug caused by controllers fighting over Instance Group management due to sligth change in logic implemented in k8s.io/ingress-gce controller.

**Comment by cezarygerard** [comment]
@nckturner 
gentle ping

**Comment by cezarygerard** [comment]
@cheftako @nckturner @bowei 
who can give approve label to this PR?

**Pull request merged**